Service/ServiceProxy.cs

```diff
@@ -0,0 +1,123 @@
+        private static readonly HttpClient Client;
+
+        static ServiceProxy()
+        {
+            Client = new HttpClient();
+        }
+
+        public static HttpRequestHeaders GetDefaultHeaders()
+        {
+            return Client.DefaultRequestHeaders;
+        }
+
+        public ServiceProxy(string serviceName, InitServiceOptions options)
+        {
+            ServiceName = serviceName;
+            Options = options;
+        }
+
+        private Uri BuildUrl(string path, string queryString, InitServiceOptions options)
+        {
+            var uriBuilder = new UriBuilder
+            {
+                Host = ServiceName,
+                Path = path,
+                Port = options?.Port ?? Options.Port,
+                Scheme = (options?.Protocol ?? Options.Protocol).ToString(),
+                Query = queryString
+            };
+            return uriBuilder.Uri;
+        }
+
+        private Dictionary<string, string> GetMergedHeaders(ServiceOptions options)
+        {
+            var result = Options.Headers ?? new Dictionary<string, string>();
+            if (options == null) return result;
+            foreach (var (key, value) in options.Headers)
+            {
+                result.Add(key, value);
+            }
+
+            return result;
+        }
+
+        private void AddRequestHeaders(HttpRequestMessage message, ServiceOptions options)
+        {
+            var mergedHeaders = GetMergedHeaders(options);
+            foreach (var (key, value) in mergedHeaders)
+            {
+                message.Headers.Remove(key);
+                message.Headers.Add(key, value);
+            }
+        }
+
+        private async Task<HttpResponseMessage> SendAsyncRequest(HttpMethod method, string path, string queryString,
+            string body, ServiceOptions options)
+        {
+            try
+            {
+                var httpRequestMessage = new HttpRequestMessage
+                {
+                    Method = method,
+                    RequestUri = BuildUrl(path, queryString, options),
+                    Content = new StringContent(body, Encoding.UTF8, MediaTypeNames.Application.Json)
+                };
+
+                AddRequestHeaders(httpRequestMessage, options);
+                var response = await Client.SendAsync(httpRequestMessage);
+                return response;
+            }
+            catch (HttpRequestException e)
+            {
+                // TODO log error
+                return new HttpResponseMessage(HttpStatusCode.InternalServerError);
+            }
+        }
+
+        public async Task<HttpResponseMessage> Get(string path, string queryString = "", string body = "",
+            ServiceOptions options = null)
+        {
+            return await SendAsyncRequest(HttpMethod.Get, path, queryString, body, options);
+        }
+
+        public async Task<HttpResponseMessage> Post(string path, string queryString = "", string body = "",
```

Comment:
Why returning a Task? Does the developer use the task methods or the `HttpResponseMessage` methods? Because reading the tests it seems like you're using methods from the second one; not the ones from the Task. How does this work?

Reply:
The `Task` class is used to wrap async methods return types. It wraps the return type of the async operation. https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task?view=netcore-3.1